core/src/main/scala/kafka/server/OffsetCheckpoint.scala

```diff
@@ -31,39 +31,45 @@ class OffsetCheckpoint(val file: File) extends Logging {
 
   def write(offsets: Map[TopicAndPartition, Long]) {
     lock synchronized {
-      // write to temp file and then swap with the existing file
-      val temp = new File(file.getAbsolutePath + ".tmp")
-
-      val fileOutputStream = new FileOutputStream(temp)
-      val writer = new BufferedWriter(new OutputStreamWriter(fileOutputStream))
       try {
-        // write the current version
-        writer.write(0.toString)
-        writer.newLine()
-      
-        // write the number of entries
-        writer.write(offsets.size.toString)
-        writer.newLine()
+        // write to temp file and then swap with the existing file
+        val temp = new File(file.getAbsolutePath + ".tmp")
+
+        val fileOutputStream = new FileOutputStream(temp)
+        val writer = new BufferedWriter(new OutputStreamWriter(fileOutputStream))
+        try {
+          // write the current version
+          writer.write(0.toString)
+          writer.newLine()
 
-        // write the entries
-        offsets.foreach { case (topicPart, offset) =>
-          writer.write("%s %d %d".format(topicPart.topic, topicPart.partition, offset))
+          // write the number of entries
+          writer.write(offsets.size.toString)
           writer.newLine()
+
+          // write the entries
+          offsets.foreach { case (topicPart, offset) =>
+            writer.write("%s %d %d".format(topicPart.topic, topicPart.partition, offset))
+            writer.newLine()
+          }
+
+          // flush the buffer and then fsync the underlying file
+          writer.flush()
+          fileOutputStream.getFD().sync()
+        } finally {
```

Comment:
In my opinion, we only do try/catch if for the part of the code that would throw the exception we intend to catch. That provides the benefit of making code more readable since, with a smaller try/catch scope, it is easier for other developers to narrow own the part of the code that may throw exception. In general it is better to use the approach with less code and less try/catch construct if the same effect is achieved.

I called it fine-granular in the sense that less code is affected by the try/catch.